contracts/Verifier.sol

```diff
@@ -0,0 +1,257 @@
+pragma solidity 0.5.2;
+pragma experimental ABIEncoderV2;
+
+import "./IEnforcer.sol";
+import "./IVerifier.sol";
+import "./IEthereumRuntime.sol";
+import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+//import "./MerkleProof.slb";
+//import "./Hash.slb";
+
+
+contract Verifier is Ownable {
+
+    struct Proofs {
+        bytes stackHash;
+        bytes memHash;
+        bytes dataHash;
+    }
+
+    struct ExecutionState {
+        bytes code;
+        bytes data;
+        uint[] stack;
+        bytes mem;
+        bytes returnData;
+        bytes logHash;
+        uint pc;
+        uint gasRemaining;
+        // accounts...
+    }
+
+    struct ComputationPath {
+        bytes32 left;
+        bytes32 right;
+    }
+
+    struct Dispute {
+        // codehash?
+        bytes32 codeHash;
+        address challengerAddr;
```

Comment:
why do we need this guy? never used, right?

Reply:
🤔 having all the data that will be presented during the dispute hashed into the root would indeed be super cool.

Reply:
Yea 😄 . But isn't going to happen, unless we restrict code execution (like jumps). 
However, we can make us of a `EXTCODECOPY`, as the code of the execution we want to verify is likely a deployed contract? (Hopes)

Reply:
> But isn't going to happen, unless we restrict code execution (like jumps).

jumps only move the program counter in the code array, no?
you actually mean internal transactions, like `CALL`, `DELEGATECALL` and so on, right?

Reply:
Unless we focus only on state changes. If so, we could do that..

Reply:
> > But isn't going to happen, unless we restrict code execution (like jumps).
> 
> jumps only move the program counter in the code array, no?

Yes. And this property makes it not possible to have everything under one 'roof'.

Reply:
💭👀 watching my brain stuttering: `but don't we commit to the code array?`

Reply:
> 
> In that case, wouldn't solver and challenger agree to operate their program counter under the same `roof` by committing to the code and its length?

we could hash the `{code, pc}` for each step (in reverse), and `shift` it to compare the `{code, pc}` submitted in `submitProof` with the `anchor hash` to validate that the code and pc is indeed the continuation from the left step and the input for the `right` step < which we want to execute onchain. However, this would lead to the fact that *if* they do not agree on the 'code path', they do end up validating the first execution step , instead of (let's say) the 20th step < there the convergence did happen 🤔

Reply:
> makes it impossible to commit linearly

let me see if i understand this. In each stateHash you have the `PC`, which is between 0 and `code.length`. And we have a code hash or deployed contract address to read the next `OPCODE` at `PC` from. 

Ok, so it is not a 'linear commitment'. and the code is not in any statehash, but still solver and challenger operate their `PC` in the same range, right?

Reply:
> that we register the address of the code for this particular execution

cool. Ideally I would have a registry contract that holds templates with all spending conditions that are useable on the plasma chain.

Reply:
Copying only the code needed given from a `contract address` is actually a very nice thing.
Or do we need the ability validating `code` which isn't somewhere stored onchain?

Reply:
> > makes it impossible to commit linearly
> 
> let me see if i understand this. In each stateHash you have the `PC`, which is between 0 and `code.length`. And we have a code hash or deployed contract address to read the next `OPCODE` at `PC` from.
> 
> Ok, so it is not a 'linear commitment'. and the code is not in any statehash, but still solver and challenger operate their `PC` in the same range, right?

Yes

Reply:
> Or do we need the ability validating code which isn't somewhere stored onchain?

for the Spending Coditions use for Plasma Leap i don't see a reason to have off-chain code, but if people want to use EVM enforcer in a broader sense, they might want to 'bring your own code'.

Reply:
*Conspiracy* 
![screenshot 2019-01-08 at 13 23 48](https://user-images.githubusercontent.com/40266861/50831615-b87a2b80-134b-11e9-8f9e-cc2383261559.png)

Reply:
<img width="416" alt="screen shot 2019-01-08 at 1 48 35 pm" src="https://user-images.githubusercontent.com/659301/50831749-2aeb0b80-134c-11e9-98eb-f850c4ae9702.png">

Reply:
Umbrella Corporation wants you 😆

Reply:
Maybe `fra-` is actually a hint for DE-CIX. Does someone punch the intertubes? 😆